### PR TITLE
[TRTLLM-13409][fix] stop reporting disagg readiness for workers that have died

### PR DESCRIPTION
**Pull request by JunyiXu-nv**
## The bug

Without a cluster manager, `DisaggCoordinatorService.is_ready()` ended in an unconditional `return True`.

That made `/health` a statement about *startup having completed*, not about the workers. Once startup finished, a ctx or gen worker could die and the coordinator kept answering **200 OK on its behalf**. A client polling `/health` had no way to learn the group was unusable, so it waited out its entire timeout against a server that could never respond.

This is distinct from a crash *during* startup — that case is connection-refused and is handled elsewhere. This one is about everything after startup succeeds.

## The information was already being collected

When a metadata server is configured, the routers already discover this and nothing consulted them:

- `Router._monitor_servers()` polls on an interval
- `check_servers_health()` filters to live servers
- a dead worker is dropped from `Router.servers`

## The fix

Key readiness off the router server lists. An empty list is unambiguous — disaggregated serving needs at least one context **and** one generation server, so zero of either cannot be served whatever the cause.

Two things in `Router` had to change before that predicate could mean anything. Both were found in review, and both were cases where the coordinator failed *open*:

**The empty state had to be reachable.** `_filter_servers_by_role()` raised `RuntimeError` on an empty live list and the monitor's `except` re-raised, so the task died with `self._servers` frozen on its last known-good value. The role could never be observed as gone — the exact case in the PR title. It now returns `[]`, and a failed poll is logged and retried instead of ending the loop.

**A monitor that has stopped keeping up must not be trusted.** Once a failed poll no longer kills the task, "the list is fresh" stops being implied by "the task is alive". `monitoring_is_stale()` reports a monitor that has ended, or that has not completed a poll within a few refresh intervals, and readiness fails closed on it. The reference point is the last successful poll, falling back to the monitor's start time — otherwise a monitor that has *never* succeeded (metadata unreachable from startup onwards, while the routers still hold the static list from the disagg config) would never age into staleness, and `/health` would promise readiness forever.

## Deliberate choices

**Not sticky.** A metadata-driven deployment adds and removes workers as a matter of course. Latching "dead" on the first removal would turn a routine topology change into a permanently unhealthy coordinator. This reports the current fact, so recovery shows up as recovery.

**A slow first poll is not staleness.** Only a poll that has not landed within the bound is. Reporting not-ready while startup is still converging would flap `/health` on every deployment — a worse failure than the one being fixed.

**Static deployments are unchanged.** With no metadata server there is no monitor, the lists never shrink, and readiness stays `True` exactly as before. That deployment shape still has no post-startup liveness signal; this PR does not claim to fix it.

**Generation-only benchmark mode is exempt from the context requirement.** `TRTLLM_DISAGG_BENCHMARK_GEN_ONLY=1` intentionally configures no context servers, so readiness requires a generation server always and a context server only when that mode is off.

**Cluster-manager path untouched.** It still delegates to `is_ready_with_router()` verbatim, with the router counts forwarded unchanged.

## Tests

`tests/unittest/others/test_disagg_readiness_after_startup.py` — no GPU, no server. Two layers, deliberately:

**`TestReadinessPredicate`** drives `is_ready()` against stub routers: ready while both roles have servers; **not** ready with no ctx, no gen, or neither (the regression — all three returned `True` before); recovery when a replacement worker arrives; the cluster manager's verdict wins with its arguments forwarded unchanged; the static deployment shape unchanged across repeated calls; generation-only mode needs no context server; stale monitoring fails closed.

**`TestMonitorDrivesReadiness`** drives a real `RoundRobinRouter` through `_monitor_servers()` with a stubbed metadata server. This layer exists because the predicate layer cannot observe whether the monitor ever *produces* the states it assumes — which is why the original revision keyed readiness off an empty list that could not occur. It asserts the empty-role transition is actually published; that the loop survives a failing poll; that a dead monitor reports stale; that a static router never does; and that a monitor which never succeeds still ages into staleness.

The file carries `pytest.mark.cpu_only` and is collected by `l0_cpu.yml` through its `unittest/others` directory entry.

## Validation status

**Readiness logic and its tests: verified on hardware.** Built from source at `c2f61a4` on an H100
node and ran the real suite — 15 passed; 15 again under `-m cpu_only`; the existing
`tests/unittest/disaggregated/test_router.py` still 86 passed; and with `router.py` reverted to
before the fail-open fix, exactly the one intended test fails and the other 14 pass. Repeat runs are
clean both idle (12/12) and under ~2x CPU oversubscription on 48 cores (12/12). Details in
[this comment](https://github.com/NVIDIA/TensorRT-LLM/pull/17206#issuecomment-5448708318).

**Still outstanding:** an end-to-end disagg deployment with a worker killed mid-benchmark. The unit
coverage above exercises the readiness predicate and the monitor that feeds it, not a real cluster.

*Note on CI collection:* the auto-generated summary below repeats that the test file has no
`test-db/` entry. It is reached through the directory entry `unittest/others` at
`tests/integration/test_lists/test-db/l0_cpu.yml:104`, combined with the file's `pytest.mark.cpu_only`
marker — confirmed by the `-m cpu_only` run above collecting all 15 tests. No per-file entry is needed.

## Dev Engineer Review

- Readiness now fails closed when required workers are unavailable or monitoring is stale.
- Router monitoring publishes empty role lists, retries failed polls, and tracks monitor freshness correctly.
- Generation-only mode does not require context workers.
- Static deployments and cluster-manager behavior remain unchanged.
- No configuration or test-list changes were found.
- Live disaggregated deployment validation remains outstanding.

## QA Engineer Review

- Added `tests/unittest/others/test_disagg_readiness_after_startup.py`.
- Added tests for readiness predicates, worker recovery, cluster-manager delegation, static deployments, generation-only mode, stale monitors, monitor recovery, empty server lists, and dead monitors.
- The test file has no entry in `tests/integration/test_lists/test-db/` or `tests/integration/test_lists/qa/`.
- Verdict: insufficient. CPU-only coverage exists, but CI test-list coverage and live disaggregated deployment validation remain outstanding.

### CODE REVIEW

File: tensorrt_llm/serve/disagg_coordinator.py [hunk 1]
```diff
@@ -324,7 +324,30 @@ async def is_ready(self) -> bool:
...
+        # Deliberately NOT sticky. A metadata-driven deployment legitimately
+        # adds and removes workers, so latching "dead" on the first removal
+        # would turn a routine topology change into a permanently unhealthy
+        # coordinator. This reports the current fact and lets recovery show up
+        # as recovery.
+        #
+        # With a static server list and no metadata server there is no monitor
+        # to consult, so the lists never shrink and this stays `True` exactly
+        # as before -- no behaviour change for that deployment shape.
+        return bool(self._ctx_router.servers) and bool(self._gen_router.servers)
```
Comment (BowenFu) [inline tensorrt_llm/serve/disagg_coordinator.py]: This predicate breaks two existing cases: `TRTLLM_DISAGG_BENCHMARK_GEN_ONLY=1` intentionally has an empty context router, so `/health` becomes permanently 503; and metadata monitoring never makes `servers` empty because `_monitor_servers()` asserts `final_servers` before assignment and exits with the stale list. Could you preserve generation-only readiness and make the monitor publish the empty-role transition, with a test that exercises the real monitor path?

Comment (brnguyen2) [inline tensorrt_llm/serve/disagg_coordinator.py]: This can't fire in the scenario the comment describes. `Router._monitor_servers()` never lets `servers` reach empty for a role: when `check_servers_health()` returns no live servers, `_filter_servers_by_role()` raises `RuntimeError("No servers available")` (`router.py:574`) — and the `assert final_servers` at `router.py:536` would fire otherwise — the `except` at `router.py:565` logs and re-`raise`s, and the monitor task dies. `self._servers` keeps the stale entries, so `is_ready()` still returns `True`.

Net effect: last-worker-of-a-role death (the case in the PR title) is not detected, and the monitor silently stops updating anything from then on. Partial death (1 of N) does shrink the list, but stays non-empty and was already fine.

To make this real, the empty state has to be allowed to exist — drop the assert/raise so the monitor can write an empty list for a role and let this check consume it — or key readiness off the health-check result directly instead of off the list. Either way, a monitor task that has died should also make the coordinator not-ready; today it fails open.

File: tensorrt_llm/serve/disagg_coordinator.py [hunk 2]
```diff
@@ -324,7 +324,30 @@ async def is_ready(self) -> bool:
                 self._ctx_router.num_prepared_servers,
                 self._gen_router.num_prepared_servers,
             )
-        return True
+        # Without a cluster manager this used to be an unconditional `True`,
```
Comment (brnguyen2) [inline tensorrt_llm/serve/disagg_coordinator.py]: 24 lines of comment for a one-line predicate, most of it PR-description narrative (what the old code did, why it's not sticky, what a static deployment does). Two lines carry the reader: what an empty list means and that it is deliberately not sticky. The rest belongs in the commit message.

File: tests/unittest/others/test_disagg_readiness_after_startup.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        return len(self._servers)
+
+
+def _coordinator(ctx_servers, gen_servers, cluster_manager=None):
+    """An DisaggCoordinatorService with only the readiness surface populated.
+
+    __init__ builds sessions and config we do not need, so the object is
+    created without running it -- the method under test reads three attributes.
+    """
+    c = object.__new__(DisaggCoordinatorService)
```
Comment (brnguyen2) [inline tests/unittest/others/test_disagg_readiness_after_startup.py:99]: `object.__new__` plus a stub exposing only `servers`/`num_prepared_servers` means these tests can only ever confirm the predicate you just wrote — they assert nothing about whether a real `Router` ever reaches the state they simulate. That's why the monitor's `assert final_servers` / re-raise path (which prevents exactly the empty-list state this file's parametrized case assumes) isn't caught here.

At minimum add a test that drives a real `Router` through `_monitor_servers()` with a stubbed metadata server returning zero healthy servers for a role, and asserts the coordinator ends up not-ready. That test fails today, which is the point.

File: tensorrt_llm/serve/disagg_coordinator.py (hunk 1 above)
Comment (JunyiXu-nv) [reply tensorrt_llm/serve/disagg_coordinator.py]: Confirmed and fixed in `269e7f8`. Your trace was exactly right, and I verified each step in the source before changing anything:

- `_filter_servers_by_role` raised `RuntimeError("No servers available")` (`router.py:575`) when the health check returned nothing
- `assert final_servers` (`:538`) would have fired otherwise
- the `except` logged and **re-raised** (`:565-567`), escaping `while True` and ending the monitor task
- `self._servers` kept its stale entries, so `is_ready()` stayed `True`

So the case in the PR title was undetectable, and the monitor stopped updating anything from that point on. It failed open.

Took your first option — make the empty state exist:

- `_filter_servers_by_role` returns `[]` instead of raising. The empty list is a legitimate observation, and it is what makes the predicate mean anything.
- A poll error no longer ends the loop; it is logged and retried, with `_last_successful_poll` deliberately left un-updated so the gap becomes visible.
- `monitoring_is_stale(max_age)` reports a monitor task that has ended or has not completed a poll within a bound. Readiness consults it and reports not-ready, so a dead monitor no longer fails open. Static server lists have no monitor and are never stale.

Blast radius was contained: `_filter_servers_by_role` has exactly one caller.

Comment (JunyiXu-nv) [reply tensorrt_llm/serve/disagg_coordinator.py]: Both confirmed and fixed in `269e7f8`.

**Generation-only.** Verified: `get_ctx_gen_server_addrs()` builds `ctx_server_urls` purely from configs with `type == "ctx"`, so a `TRTLLM_DISAGG_BENCHMARK_GEN_ONLY=1` deployment genuinely has an empty context router and my predicate would have pinned `/health` at 503. Readiness now requires a generation server always, and a context server only when that mode is off. The env read is a shared `gen_only_benchmark_enabled()` helper rather than duplicated inline — that also removed a redundant local `import os` in `_wait_for_all_servers_ready`.

Covered by `test_generation_only_benchmark_needs_no_context_server`, which also pins that a generation worker is still mandatory in that mode.

**Monitor publishing the empty transition.** This was the same root cause brnguyen2 traced: `_filter_servers_by_role` raised on an empty live list and the monitor's `except` re-raised, so `servers` never reached empty for a role and the task died holding a stale list. It now returns `[]`, the loop survives poll errors, and `monitoring_is_stale()` lets readiness fail closed when monitoring has stopped.

**Test through the real monitor**, as you asked. `TestMonitorDrivesReadiness` drives a real `RoundRobinRouter` through `_monitor_servers()` with a stubbed metadata server and asserts the empty-role transition is actually published, the loop survives a failing poll, and a dead monitor reports stale. Verified against a build on an H100 — see the reply on the test thread for the before/after numbers.

File: tests/unittest/others/test_disagg_readiness_after_startup.py (hunk 3 above)
Comment (JunyiXu-nv) [reply tests/unittest/others/test_disagg_readiness_after_startup.py:99]: Fixed in `269e7f8`, and you were right that this was the reason the assert/raise problem went unseen.

Added `TestMonitorDrivesReadiness`, which builds a real `RoundRobinRouter` with a stubbed metadata server and a stubbed `_check_server_health`, then runs `_monitor_servers()` for real:

- `test_monitor_publishes_empty_list_when_role_dies` — every worker of the role is unhealthy; asserts `servers == []` **and** that the monitor task is still running
- `test_monitor_survives_a_failing_poll` — a transient `fetch_live_servers` error; asserts the task is alive and polled again
- `test_dead_monitor_is_reported_stale` — `monitoring_is_stale()` is what lets readiness fail closed
- `test_static_router_is_never_stale` — no monitor means a static list, which cannot go stale

Verified on an H100 build that these genuinely fail without the fix, by reverting all three router guards (the `RuntimeError`, the `assert`, and the `raise`):

| router | result |
| --- | --- |
| with fix | **14 passed** |
| all three guards reverted | **2 failed**, 12 passed |

The two failures are `test_monitor_publishes_empty_list_when_role_dies` and `test_monitor_survives_a_failing_poll`.

Worth noting the other **12 still passed** in that run — every stub-router predicate test included. That is your point made concrete: the predicate layer cannot observe the monitor's behaviour, so it could never have caught this. I kept those tests (they are fast and do cover the predicate's own branches) but they are now explicitly the shallower of two layers, and the file docstring says so.

File: tensorrt_llm/serve/disagg_coordinator.py (hunk 2 above)
Comment (JunyiXu-nv) [reply tensorrt_llm/serve/disagg_coordinator.py]: Fixed in `269e7f8`. Trimmed from 24 lines to the two that carry the reader — what an empty list means and that it is deliberately not sticky — with the narrative moved into the commit message where you suggested.

The predicate did grow past one line, but only because it now has real branches to explain: the static-list short-circuit, the staleness check, and the generation-only exemption. Each has a sentence saying why it exists, rather than recounting what the old code did.

File: tensorrt_llm/serve/router.py [hunk 4]
```diff
@@ -378,6 +382,26 @@ def servers(self) -> List[str]:
...
+        has no monitor to go stale, and its list is correct by construction.
+        """
+        if self._monitor_task is None:
+            return False
+        if self._monitor_task.done():
+            return True
+        if self._last_successful_poll is None:
+            # Started, but the first poll has not landed yet. Treat the task's
+            # own start as the reference point rather than reporting stale.
+            return False
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/serve/router.py:418]: +1

[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tensorrt_llm/serve/router.py:418]: Confirmed and fixed in `12718a8`. @BowenFu — this was real, and the reachability chain is worth spelling out, because it is not visible from the diff alone:

- `build_disagg_routers()` seeds each router with the static server list from the disagg config **whether or not** a metadata server is configured (`get_ctx_gen_server_addrs(config.server_configs)` → `build_disagg_routers(...)`), so `Router._servers` starts non-empty.
- `_wait_for_all_servers_ready()` is satisfied by that list, so startup completes.
- Now that a failed poll no longer ends the loop (previous round of review), the monitor stays alive and fails every poll.
- `_last_successful_poll` therefore stays `None`, `monitoring_is_stale()` returned `False` on that branch unconditionally, and readiness kept trusting a list nothing had ever refreshed.

Net: exactly the fail-open this PR set out to close, entered from startup rather than from a worker dying later. My previous round only shut the second door.

**Fix.** `start_server_monitoring()` now records `_monitor_started_at`, and `monitoring_is_stale()` uses it as the reference point until the first poll lands. Both checks are explicit `is None` rather than `or` — `time.monotonic()` may legitimately return `0.0`.

**Deliberately kept:** a first poll that is merely *slow* but still inside the bound stays ready. Reporting not-ready there would flap `/health` on every startup, which is a worse failure than the one being fixed. The new test pins both directions, so a future tightening cannot silently take that away.

**Test.** `test_monitoring_that_never_succeeds_ages_into_staleness` drives the real `start_server_monitoring()` with a `fetch_live_servers` that always raises, then asserts the fail-open preconditions actually hold (task alive, never succeeded, still holding the initial list) and that the coordinator reports not-ready past the bound but ready within it.

**Verification caveat, stated plainly.** I have no local build, so I validated by extracting the exact shipped source of `monitoring_is_stale()` and `is_ready()` and exercising it against stubs, rather than by running pytest. Running the same 12 assertions against pre-fix `HEAD`:

| assertion | pre-fix | post-fix |
| --- | --- | --- |
| never-succeeded monitor past the bound → stale | **FAIL** (got `False`) | PASS |
| same state → `is_ready()` | **FAIL (got `True` — the fail-open)** | PASS |
| other 10 regression guards | PASS | PASS |

Only the two intended assertions flip, which is the evidence that the change is narrow. The pytest file itself still needs a run in a real environment before merge, and I am not claiming otherwise.

**On the type annotations in the same comment:** `_filter_servers_by_role()` is annotated in `fe60a8b` — its return type is the load-bearing part of this branch (it returns `[]` where it used to raise), so the signature should say so. I skipped the blanket `-> None` on the test methods: that is not the convention anywhere else under `tests/unittest/`, and adding it in this one file would be noise rather than consistency.

File: tensorrt_llm/serve/router.py [hunk 5]
```diff
@@ -532,10 +556,15 @@ async def _monitor_servers(self, poll_interval: float = 10.0):
...
+                # state -- every worker of this role is gone -- and must be
+                # published so readiness reflects it. Asserting here instead
+                # would kill the loop and leave a stale, healthy-looking list.
                 final_servers = role_specific_servers
-
-                assert final_servers, f"No {self._server_role} servers available"
+                if not final_servers:
+                    logger.warning(
+                        f"No live {self._server_role} servers; publishing an "
+                        "empty server list for this role")
```
[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tensorrt_llm/serve/router.py:587]: Checked this against the code. It is accurate as far as it goes, but I would like to leave it out of this PR, for three reasons.

**It only fires when every worker of every role is gone.** `fetch_live_servers()` scans all `trtllm/` keys, not this router's role. If the ctx workers all deregister while a gen worker remains, `server_key_map` is non-empty, no `ValueError` is raised, and the empty-role transition flows through `_filter_servers_by_role()` returning `[]` — the path this PR added. So the scenario here is specifically "the entire cluster deregistered", not "a role died".

**In that scenario readiness already fails closed**, via the staleness path rather than the empty-list path: no poll succeeds, `_last_successful_poll` stops advancing, `monitoring_is_stale()` trips, `/health` reports not-ready. What the change would buy is latency — one poll interval instead of the staleness bound — not correctness.

**And it carries a contract change this PR should not absorb.** `test_fetch_live_servers_context` (`tests/unittest/disaggregated/test_router.py:1537`) explicitly asserts `pytest.raises(ValueError)` for the empty-metadata read, parametrized across all three router classes. Changing the return contract means rewriting those assertions — a separate concern from "stop reporting readiness for dead workers" (see the one-concern-per-PR rule in `AGENTS.md`).

For the record I agree with the underlying point: a successful metadata read that returns zero workers is a legitimate observation, not an error — the same argument that made me change `_filter_servers_by_role()`. It deserves a follow-up that also updates the existing contract test. Just not here.

File: tensorrt_llm/serve/router.py [hunk 6]
```diff
@@ -562,17 +591,32 @@ async def _monitor_servers(self, poll_interval: float = 10.0):
...
+                self._last_successful_poll = time.monotonic()
+            except asyncio.CancelledError:
+                raise
             except Exception as e:
+                # Keep polling. Re-raising ends the loop, and a monitor that is
+                # no longer running leaves ``self._servers`` frozen on its last
+                # value -- which readiness would then read as healthy forever.
+                # ``_last_successful_poll`` is deliberately not updated, so
+                # ``monitoring_is_stale()`` starts reporting the gap.
                 logger.error(f"Error in server monitoring: {e}")
```
[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tensorrt_llm/serve/router.py:623]: Partly pushing back on this one.

**The breadth is pre-existing.** This handler was already `except Exception as e:` before the PR; the diff removed the `raise`, it did not widen the catch. Narrowing it now would be a separate cleanup of existing code.

**Both designs fail closed, so the safety argument is symmetric.** With a broad catch, an unexpected error means the task stays alive and never completes a poll → `monitoring_is_stale()` trips → not-ready. With a narrow catch, an unexpected error kills the task → `_monitor_task.done()` → `monitoring_is_stale()` trips → not-ready. Either way the failure surfaces at `/health`; nothing is hidden by the breadth. That was the point of adding staleness alongside the swallow, rather than instead of it.

**Narrowing has a downside the current shape does not.** The metadata backend is pluggable (etcd, consul, ...), and the transport errors reachable from `fetch_live_servers()` / `check_servers_health()` are not enumerable from this call site. Guessing the set wrong turns a transient backend hiccup into a permanently dead monitor — the exact regression brnguyen2 asked me to fix two rounds ago. The asymmetry matters: guessing too narrow breaks liveness, guessing too broad costs a delayed report.

**On the lint citation:** `BLE001` is not in this repository's rule set — `[tool.ruff.lint] select = ["D", "E", "F", "I", "PLE", "W"]` in `pyproject.toml` — and `tensorrt_llm/serve/router.py` is listed in `legacy-files.txt`, so it goes through the baseline-gated `ruff-legacy` hook rather than the main `ruff` hook. So this is a guideline question, not a CI failure.

Happy to concede the debuggability half of it: I can add `exc_info` to the `logger.error` so the traceback survives rather than just the message, which addresses "hides the original failure behind repeated logs" without trading away liveness. Say the word and I will push it.

**Review by BowenFu — APPROVED** [review]

File: tests/unittest/others/test_disagg_readiness_after_startup.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        list, the monitor's ``except`` re-raised, the task died, and
+        ``servers`` kept its stale entries -- so this assertion failed.
+        """
+
+        async def scenario():
+            router = self._router(["ctx0"], healthy=set())  # worker is dead
+            # One poll, then stop: _monitor_servers loops forever by design.
+            task = asyncio.create_task(router._monitor_servers(poll_interval=0.01))
+            router._monitor_task = task
+            await asyncio.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tests/unittest/others/test_disagg_readiness_after_startup.py]: Good catch, and more right than the wording suggests — fixed in `c2f61a4`. Verifying it turned up two *different* failure modes, not one:

**The two pre-existing tests could fail for the wrong reason.** `test_monitor_publishes_empty_list_when_role_dies` and `test_monitor_survives_a_failing_poll` assert *positive progress* — a published list, a second poll — inside `sleep(0.1)` / `sleep(0.15)`. Starve the event loop and they go red with nothing wrong in the code.

**My own test failed the opposite way, and this is the part I had missed.** `test_monitoring_that_never_succeeds_ages_into_staleness` asserts `_last_successful_poll is None`, which is *equally true* of a monitor that polled and failed and one that was never scheduled at all. The staleness assertion is wall-clock so it never flaked — it just could have gone green without `always_fails()` ever running. Your phrasing ("also passes without proving that always_fails() ran") named that exactly.

What changed:

- Each stubbed operation signals an `asyncio.Event`; assertions wait on it via `asyncio.wait_for` with a generous bound that is a deadlock guard, not a timing assertion.
- The empty-list case hooks **`_on_servers_updated`** rather than the fetch. It runs under the monitor's lock *after* `self._servers` is reassigned, so the assertion reads a published list instead of racing the publish. Signalling around the fetch would have reintroduced the race in a quieter form.
- The staleness case **rewinds `_monitor_started_at`** instead of waiting out a real bound — 60s of failing polls against a 30s bound — which takes wall-clock out of the arithmetic entirely, and counts failed polls so the retry is proven rather than assumed.

Verified with the same harness I described on the fail-open thread (extracted shipped source of `_monitor_servers`, `start_server_monitoring`, `monitoring_is_stale`, `is_ready` bound onto a minimal router). To show the timing dependence is actually gone rather than just smaller, I re-ran with `poll_interval` raised 50× to 0.5s — well past what the old fixed sleeps budgeted for:

| | fast loop (0.01s) | slow loop (0.5s) |
| --- | --- | --- |
| old, fixed `sleep(0.15)` | PASS | **FAIL** — `polled again: got False` |
| new, `asyncio.Event` | PASS | PASS |

That FAIL is the CI flake you predicted, reproduced deterministically. Standard caveat from the earlier threads still applies: no local build, so this is the extracted-source harness rather than a pytest run.

File: tests/unittest/others/test_disagg_readiness_after_startup.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+
+
+def _coordinator(
+    ctx_servers,
+    gen_servers,
+    cluster_manager=None,
+    metadata_server=object(),
+    ctx_stale=False,
+    gen_stale=False,
+):
```
[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tests/unittest/others/test_disagg_readiness_after_startup.py]: Applied in `c2f61a4`, with one correction to the reasoning.

This is not the mutable-default-argument bug that B008 usually catches. `object()` is evaluated once at definition time, which is precisely what a sentinel wants — the old code was correct, not accidentally correct. And for the record B008 is not in this repository's rule set (`[tool.ruff.lint] select = ["D", "E", "F", "I", "PLE", "W"]`), so nothing was failing.

Taking it anyway because the *readability* argument holds independently: `_DEFAULT_METADATA_SERVER` at module level says "stable sentinel" out loud, where `object()` in a signature makes a reader stop and work out whether it is a per-call or per-module instance. Two lines for that is a good trade. I added a comment saying which of the two it is, so the next reader does not have to re-derive it either.

File: tests/unittest/others/test_disagg_readiness_after_startup.py [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    c._ctx_router = _StubRouter(ctx_servers, stale=ctx_stale)
+    c._gen_router = _StubRouter(gen_servers, stale=gen_stale)
+    c._disagg_cluster_manager = cluster_manager
+    c._metadata_server = metadata_server
+    c._monitor_staleness_secs = 30.0
+    return c
+
+
+def _ready(c):
+    return asyncio.run(c.is_ready())
```
[the remark this answers is not in the document]
Comment (JunyiXu-nv) [reply tests/unittest/others/test_disagg_readiness_after_startup.py]: Partly applied in `c2f61a4`, and I owe you a correction on my earlier reply to the sibling thread — I claimed annotating tests "is not the convention" from memory rather than from measurement. I measured it this time.

**You are right that the guideline is unqualified.** `CODING_GUIDELINES.md:664` says "Always annotate functions. Make the return type `None` if the function does not return anything", with no exemption for tests. My earlier reply should have cited that line and argued against it explicitly instead of gesturing at convention.

**The convention claim does hold, though.** Across all 930 `test_*.py` files under `tests/unittest/`:

| | with a return annotation |
| --- | --- |
| 11298 `test_*` functions | **10.1%** |
| 7922 helper functions | **28.7%** |

So blanket annotation is neither the written-rule status quo nor close to it. Relevant too: `[tool.mypy] files` is an explicit opt-in list covering only the sampler subtree — neither this file nor `serve/router.py` is type-checked, so annotations here are documentation, not enforcement.

**Where I landed.** Annotated everything that is a real signature someone reads or calls across a boundary — `_StubRouter`, `_coordinator`, `_ready`, `_StubMetadataServer`, `_router`, and the two new helpers `_await_progress` / `_run_monitored`. Skipped `-> None` on the `test_*` methods and the nested `scenario()` / hook closures: pytest calls those, nothing else does, the return is `None` by construction, and adding it in one file out of 930 buys inconsistency rather than consistency.

Took your "precise built-in generic types" point: this file uses `list[str]` / `dict[str, dict[str, str]]` throughout. Worth flagging the asymmetry deliberately — `_filter_servers_by_role()` in `router.py` is annotated `List[str]` / `Dict[str, str]` instead. `CODING_GUIDELINES.md:674` does prefer the builtins, but that file annotates every one of its existing signatures with `List` / `Dict`, and mixing both spellings in one module is worse than following either. New file gets the modern spelling; the legacy file stays internally consistent.

If a maintainer would rather have the blanket `-> None` everywhere, say so and I will add it — it is mechanical. I just do not think it earns its diff here.

**Comment by JunyiXu-nv** [comment]
## Hardware validation — retracting the "not run under pytest" caveat

Earlier replies on this PR carried a caveat: I had no local build, so the tests were validated by extracting the shipped source of `monitoring_is_stale()` / `is_ready()` / `_monitor_servers()` and exercising it against stubs, rather than by running pytest. **That caveat is now withdrawn.** Built from source on an H100 node and ran the real suite.

Built at `c2f61a4` — `pip show tensorrt_llm` reports `1.3.0rc25+c2f61a4945`, and the import resolves inside the branch worktree, so this is the PR's code and not `main`.

| # | check | result |
| --- | --- | --- |
| 1 | `tests/unittest/others/test_disagg_readiness_after_startup.py` | **15 passed** |
| 2 | same file under `-m cpu_only` (the path `l0_cpu` actually collects through) | **15 passed** |
| 3 | existing `tests/unittest/disaggregated/test_router.py` | **86 passed** |
| 4 | **`router.py` reverted to pre-fail-open-fix (`93a5d89`), tests kept** | **1 failed, 14 passed** |
| 5 | 12 consecutive runs, idle machine | **12/12 clean** |
| 6 | 12 consecutive runs under 96-way CPU contention on 48 cores (load avg 65.8) | **12/12 clean** |

Two of these are the ones worth reading.

**#4 is the load-bearing one.** With the fail-open fix reverted but the new tests kept, exactly one test fails — `test_monitoring_that_never_succeeds_ages_into_staleness` — and the other 14 still pass. The test locks down the specific bug rather than riding along. (Working tree restored afterwards; no drift.)

**#6 answers the flakiness finding directly.** The concern was that fixed `sleep()` calls would not survive a loaded CI worker. Oversubscribing a 48-core box ~2× and re-running 12 times produced no failures. For contrast, before the `asyncio.Event` change, simply raising `poll_interval` past what the old sleeps budgeted for reproduced the failure deterministically (`polled again: got False`) — so the timing dependence was real, and it is now gone rather than merely narrowed.

**#3 also settles a claim I made** on the `fetch_live_servers()` thread: I said I had not touched that `ValueError` contract, and `test_fetch_live_servers_context` (which asserts it, parametrized across all three router classes) still passes — along with the other 85.

### Still not done

This validates the readiness logic and its tests. It is **not** an end-to-end disagg run with a worker killed mid-benchmark — that remains outstanding, and the PR description says so. If a reviewer considers it required before merge rather than nice-to-have, say so and I will set one up.

<details>
<summary>Environment</summary>

- Node: `ipp2-0100`, H100 80GB, 48 cores, 256 GB
- Container: `tensorrt_llm/devel:latest` via `make -C docker devel_build`
- Build: `python3 scripts/build_wheel.py -G Ninja -a 90-real -f`, ~48 min cold, exit 0
- pytest 9.0.3

One build gotcha worth recording for anyone reproducing this: `build_wheel.py -f` does not clear the CMake cache. A `cpp/build/CMakeCache.txt` written at a different absolute path (host path vs. the container's `/code/tensorrt_llm`) fails the build with `ninja: error: mkdir(...): Permission denied`. `rm -rf cpp/build` first.

</details>

**Review by BowenFu — APPROVED** [review]

**Review by asfiyab-nvidia — APPROVED** [review]

File: tensorrt_llm/serve/disagg_coordinator.py
Comment (BowenFu) [inline tensorrt_llm/serve/disagg_coordinator.py]: Thanks for putting this together! The 1.3 release branch is about to be cut from main (planned for 9/29), and until then we're asking larger changes and refactors like this one to wait so main stays stable for the branch-out. If this really needs to make 1.3 or a nightly/weekly release, could you explain why in the PR description and in your review request? I'll resolve this thread once that's there, or after branch-out either way. This is only a timing hold, not a concern with the change itself.
